File: backend/src/agents/assessment.agent.ts
import callGranite from "../services/granite.service";

interface AssessmentInput {
  name: string;
  careerGoal: string;
  currentSkill: string;
  weeklyHours: number;
  learningStyle: string;
  preferredLanguage: string;
  interests: string[];
}

function extractJson(text: string) {
  try {
    return JSON.parse(text.trim());
  } catch {
    const start = text.indexOf("{");
    const end = text.lastIndexOf("}");
    if (start !== -1 && end !== -1) {
      const jsonStr = text.substring(start, end + 1);
      try {
        return JSON.parse(jsonStr);
      } catch (e) {
        throw new Error(`Failed to parse extracted JSON string: ${e}`);
      }
    }
    throw new Error("No JSON object found in response");
  }
}

export async function assessStudent(data: AssessmentInput) {
  const {
    name,
    careerGoal,
    currentSkill,
    weeklyHours,
    learningStyle,
    preferredLanguage,
    interests,
  } = data;

  const prompt = `
You are an expert AI Career Mentor.

Analyze the student's profile and provide a personalized assessment.

Student Profile

Name: ${name}
Career Goal: ${careerGoal}
Current Skill: ${currentSkill}
Weekly Study Hours: ${weeklyHours}
Learning Style: ${learningStyle}
Preferred Language: ${preferredLanguage}
Interests: ${interests.join(", ")}

Respond ONLY in JSON using this exact format:

{
  "skillScore": number,
  "difficulty": "...",
  "knowledgeGaps": ["...", "..."],
  "strengths": ["...", "..."],
  "estimatedDuration": "...",
  "personalizedAdvice": "..."
}
`;

  try {
    const response = await callGranite(prompt);

    console.log("========== GRANITE RESPONSE ==========");
    console.log(response.data.text);
    console.log("======================================");

    const parsed = extractJson(response.data.text);
    return parsed;
  } catch (error) {
    console.error("=================================");
    console.error("IBM GRANITE ERROR / PARSE ERROR");
    console.error(error);
    console.error("=================================");

    // Fallback in case Granite is offline or returns invalid format
    return {
      skillScore: currentSkill === "Beginner" ? 30 : currentSkill === "Intermediate" ? 60 : 85,
      difficulty: currentSkill,
      knowledgeGaps: [
        `Core ${careerGoal} concepts`,
        "Enterprise architecture and tools",
      ],
      strengths: [
        "Commitment to weekly study hours",
        "Clear professional career objective",
      ],
      estimatedDuration: currentSkill === "Beginner" ? "6 Months" : currentSkill === "Intermediate" ? "4 Months" : "2 Months",
      personalizedAdvice: `Based on your interest in ${interests.join(", ")}, we recommend focusing on hands-on practical building using a ${learningStyle} learning style.`,
    };
  }
}

File: backend/agents/assessment_agent.py
import json
from services.granite_service import call_granite

def extract_json(text: str) -> dict:
    text = text.strip()
    try:
        return json.loads(text)
    except Exception:
        start = text.find("{")
        end = text.rfind("}")
        if start != -1 and end != -1:
            json_str = text[start:end + 1]
            try:
                return json.loads(json_str)
            except Exception as e:
                raise ValueError(f"Failed to parse extracted JSON string: {e}")
        raise ValueError("No JSON object found in response")

def assess_student(data: dict) -> dict:
    name = data.get("name", "Student")
    career_goal = data.get("careerGoal", "Frontend Developer")
    current_skill = data.get("currentSkill", "Beginner")
    weekly_hours = data.get("weeklyHours", 5)
    learning_style = data.get("learningStyle", "practical")
    preferred_language = data.get("preferredLanguage", "English")
    interests = data.get("interests", [])

    prompt = f"""
You are an expert AI Career Mentor.

Analyze the student's profile and provide a personalized assessment.

Student Profile

Name: {name}
Career Goal: {career_goal}
Current Skill: {current_skill}
Weekly Study Hours: {weekly_hours}
Learning Style: {learning_style}
Preferred Language: {preferred_language}
Interests: {", ".join(interests)}

Respond ONLY in JSON using this exact format:

{{
  "skillScore": number,
  "difficulty": "...",
  "knowledgeGaps": ["...", "..."],
  "strengths": ["...", "..."],
  "estimatedDuration": "...",
  "personalizedAdvice": "..."
}}
"""

    try:
        response = call_granite(prompt)
        text_out = response["data"]["text"]
        print("========== GRANITE RESPONSE ==========")
        print(text_out)
        print("======================================")
        return extract_json(text_out)
    except Exception as error:
        print("=================================")
        print("IBM GRANITE ERROR / PARSE ERROR")
        print(error)
        print("=================================")

        # Fallback in case Granite is offline or returns invalid format
        score = 30 if current_skill == "Beginner" else (60 if current_skill == "Intermediate" else 85)
        duration = "6 Months" if current_skill == "Beginner" else ("4 Months" if current_skill == "Intermediate" else "2 Months")

        return {
            "skillScore": score,
            "difficulty": current_skill,
            "knowledgeGaps": [
                f"Core {career_goal} concepts",
                "Enterprise architecture and tools"
            ],
            "strengths": [
                "Commitment to weekly study hours",
                "Clear professional career objective"
            ],
            "estimatedDuration": duration,
            "personalizedAdvice": f"Based on your interest in {', '.join(interests)}, we recommend focusing on hands-on practical building using a {learning_style} learning style."
        }
